File: src/core/exchange.ts
import * as ccxt from 'ccxt';
import { logger } from '../utils/logger';

export interface ExchangeConfig {
  apiKey: string;
  apiSecret: string;
  apiPassphrase?: string;
  testnet?: boolean;
}

export interface Balance {
  total: number;
  available: number;
  used: number;
  currency: string;
}

export interface MarketData {
  symbol: string;
  price: number;
  change24h: number;
  changePercent24h: number;
  volume: number;
  volume24h: number;
  high24h: number;
  low24h: number;
  bid?: number;
  ask?: number;
  timestamp: string;
}

export interface Ticker {
  symbol: string;
  last: number;
  bid: number;
  ask: number;
  high: number;
  low: number;
  volume: number;
  percentage: number;
  timestamp: number;
}

export class ExchangeManager {
  private exchange: ccxt.Exchange;
  private isConnected: boolean = false;

  constructor(config: ExchangeConfig) {
    const options: any = {
      apiKey: config.apiKey,
      secret: config.apiSecret,
      password: config.apiPassphrase,
      enableRateLimit: true,
      options: {
        defaultType: 'spot', // spot trading
      }
    };

    // Testnet support
    if (config.testnet) {
      options.urls = {
        api: {
          public: 'https://api-sandbox.kucoin.com',
          private: 'https://api-sandbox.kucoin.com',
        }
      };
    }

    this.exchange = new ccxt.kucoin(options);
    logger.info('Exchange Manager initialized for KuCoin');
  }

  /**
   * Проверка подключения к бирже
   */
  async connect(): Promise<boolean> {
    try {
      await this.exchange.loadMarkets();
      this.isConnected = true;
      logger.info('✅ Connected to KuCoin successfully');
      return true;
    } catch (error) {
      logger.error('❌ Failed to connect to KuCoin:', error);
      this.isConnected = false;
      return false;
    }
  }

  /**
   * Получить баланс USDT
   */
  async getBalance(currency: string = 'USDT'): Promise<Balance> {
    try {
      const balance = await this.exchange.fetchBalance();
      const currencyBalance = balance[currency] || { total: 0, free: 0, used: 0 };

      return {
        total: currencyBalance.total || 0,
        available: currencyBalance.free || 0,
        used: currencyBalance.used || 0,
        currency
      };
    } catch (error) {
      logger.error(`Failed to fetch balance for ${currency}:`, error);
      throw error;
    }
  }

  /**
   * Получить текущую цену тикера
   */
  async getTicker(symbol: string = 'BTC/USDT'): Promise<Ticker> {
    try {
      const ticker = await this.exchange.fetchTicker(symbol);

      return {
        symbol: ticker.symbol,
        last: ticker.last || 0,
        bid: ticker.bid || 0,
        ask: ticker.ask || 0,
        high: ticker.high || 0,
        low: ticker.low || 0,
        volume: ticker.baseVolume || 0,
        percentage: ticker.percentage || 0,
        timestamp: ticker.timestamp || Date.now()
      };
    } catch (error) {
      logger.error(`Failed to fetch ticker for ${symbol}:`, error);
      throw error;
    }
  }

  /**
   * Получить полные рыночные данные
   */
  async getMarketData(symbol: string = 'BTC/USDT'): Promise<MarketData> {
    try {
      const ticker = await this.getTicker(symbol);

      return {
        symbol: ticker.symbol,
        price: ticker.last,
        change24h: ticker.percentage,
        changePercent24h: ticker.percentage,
        volume: ticker.volume,
        volume24h: ticker.volume,
        high24h: ticker.high,
        low24h: ticker.low,
        bid: ticker.bid,
        ask: ticker.ask,
        timestamp: new Date(ticker.timestamp).toISOString()
      };
    } catch (error) {
      logger.error(`Failed to fetch market data for ${symbol}:`, error);
      throw error;
    }
  }

  /**
   * Получить OHLCV данные для анализа
   */
  async getOHLCV(
    symbol: string = 'BTC/USDT',
    timeframe: string = '1h',
    limit: number = 100
  ): Promise<any[]> {
    try {
      const ohlcv = await this.exchange.fetchOHLCV(symbol, timeframe, undefined, limit);

      return ohlcv.map((candle: any) => ({
        timestamp: candle[0],
        open: candle[1],
        high: candle[2],
        low: candle[3],
        close: candle[4],
        volume: candle[5]
      }));
    } catch (error) {
      logger.error(`Failed to fetch OHLCV for ${symbol}:`, error);
      throw error;
    }
  }

  /**
   * Создать лимитный ордер
   */
  async createLimitOrder(
    symbol: string,
    side: 'buy' | 'sell',
    amount: number,
    price: number
  ): Promise<any> {
    try {
      logger.info(`Creating ${side} limit order: ${amount} ${symbol} @ ${price}`);

      const order = await this.exchange.createLimitOrder(
        symbol,
        side,
        amount,
        price
      );

      logger.info(`Order created successfully: ${order.id}`);
      return order;
    } catch (error) {
      logger.error(`Failed to create ${side} order:`, error);
      throw error;
    }
  }

  /**
   * Создать рыночный ордер
   */
  async createMarketOrder(
    symbol: string,
    side: 'buy' | 'sell',
    amount: number
  ): Promise<any> {
    try {
      logger.info(`Creating ${side} market order: ${amount} ${symbol}`);

      const order = await this.exchange.createMarketOrder(
        symbol,
        side,
        amount
      );

      logger.info(`Market order created successfully: ${order.id}`);
      return order;
    } catch (error) {
      logger.error(`Failed to create ${side} market order:`, error);
      throw error;
    }
  }

  /**
   * Получить открытые ордера
   */
  async getOpenOrders(symbol?: string): Promise<any[]> {
    try {
      const orders = await this.exchange.fetchOpenOrders(symbol);
      return orders;
    } catch (error) {
      logger.error('Failed to fetch open orders:', error);
      throw error;
    }
  }

  /**
   * Отменить ордер
   */
  async cancelOrder(orderId: string, symbol: string): Promise<any> {
    try {
      logger.info(`Cancelling order ${orderId} for ${symbol}`);
      const result = await this.exchange.cancelOrder(orderId, symbol);
      logger.info(`Order ${orderId} cancelled successfully`);
      return result;
    } catch (error) {
      logger.error(`Failed to cancel order ${orderId}:`, error);
      throw error;
    }
  }

  /**
   * Проверить, подключен ли exchange
   */
  isExchangeConnected(): boolean {
    return this.isConnected;
  }

  /**
   * Получить минимальный размер ордера
   */
  async getMinAmount(symbol: string): Promise<number> {
    try {
      await this.exchange.loadMarkets();
      const market = this.exchange.market(symbol);
      return market.limits.amount?.min || 0.0001;
    } catch (error) {
      logger.error(`Failed to get min amount for ${symbol}:`, error);
      return 0.0001; // Fallback
    }
  }

  /**
   * Получить информацию о рынке
   */
  async getMarketInfo(symbol: string): Promise<any> {
    try {
      await this.exchange.loadMarkets();
      return this.exchange.market(symbol);
    } catch (error) {
      logger.error(`Failed to get market info for ${symbol}:`, error);
      throw error;
    }
  }
}

File: backend_py/core/exchange.py
import ccxt
from typing import Dict, Any, Optional
from loguru import logger
from pydantic import BaseModel

class ExchangeConfig(BaseModel):
    apiKey: str
    apiSecret: str
    apiPassphrase: Optional[str] = None
    testnet: bool = False

class Balance(BaseModel):
    total: float
    available: float
    used: float
    currency: str

class MarketData(BaseModel):
    symbol: str
    price: float
    change24h: float
    changePercent24h: float
    volume: float
    volume24h: float
    high24h: float
    low24h: float
    bid: Optional[float] = None
    ask: Optional[float] = None
    timestamp: str

class Ticker(BaseModel):
    symbol: str
    last: float
    bid: float
    ask: float
    high: float
    low: float
    volume: float
    percentage: float
    timestamp: int

class ExchangeManager:
    def __init__(self, config: Dict[str, Any]):
        self.config = ExchangeConfig(**config)
        self.exchange = ccxt.kucoin({
            'apiKey': self.config.apiKey,
            'secret': self.config.apiSecret,
            'password': self.config.apiPassphrase,
            'sandbox': self.config.testnet,
            'enableRateLimit': True,
        })
        self.is_connected = False

    async def connect(self) -> bool:
        try:
            # Проверяем подключение через получение баланса
            await self.exchange.loadMarkets()
            self.is_connected = True
            logger.info('Exchange Manager initialized for KuCoin')
            return True
        except Exception as e:
            logger.error(f'Failed to connect to KuCoin: {e}')
            self.is_connected = False
            return False

    async def disconnect(self):
        self.is_connected = False
        logger.info('Exchange Manager disconnected')

    async def get_balance(self, currency: str = 'USDT') -> Balance:
        try:
            balance = await self.exchange.fetchBalance()
            return Balance(
                total=balance.get(currency, {}).get('total', 0),
                available=balance.get(currency, {}).get('free', 0),
                used=balance.get(currency, {}).get('used', 0),
                currency=currency
            )
        except Exception as e:
            logger.error(f'Failed to get balance: {e}')
            return Balance(total=0, available=0, used=0, currency=currency)

    async def get_ticker(self, symbol: str) -> Optional[Ticker]:
        try:
            ticker = await self.exchange.fetchTicker(symbol)
            return Ticker(
                symbol=ticker['symbol'],
                last=ticker['last'],
                bid=ticker['bid'],
                ask=ticker['ask'],
                high=ticker['high'],
                low=ticker['low'],
                volume=ticker['baseVolume'],
                percentage=ticker['percentage'],
                timestamp=ticker['timestamp']
            )
        except Exception as e:
            logger.error(f'Failed to get ticker for {symbol}: {e}')
            return None

    async def get_market_data(self, symbol: str) -> Optional[MarketData]:
        try:
            ticker = await self.get_ticker(symbol)
            if not ticker:
                return None

            # Получаем дополнительные данные
            ohlcv = await self.exchange.fetchOHLCV(symbol, '1d', limit=2)
            if len(ohlcv) < 2:
                return None

            prev_close = ohlcv[-2][4]  # Предыдущая закрытая цена
            change24h = ticker.last - prev_close
            changePercent24h = (change24h / prev_close) * 100 if prev_close > 0 else 0

            return MarketData(
                symbol=symbol,
                price=ticker.last,
                change24h=change24h,
                changePercent24h=changePercent24h,
                volume=ticker.volume,
                volume24h=ticker.volume,
                high24h=ticker.high,
                low24h=ticker.low,
                bid=ticker.bid,
                ask=ticker.ask,
                timestamp=str(ticker.timestamp)
            )
        except Exception as e:
            logger.error(f'Failed to get market data for {symbol}: {e}')
            return None

    async def place_order(self, symbol: str, side: str, amount: float, price: Optional[float] = None, order_type: str = 'limit') -> Dict[str, Any]:
        try:
            if order_type == 'limit' and price:
                order = await self.exchange.createOrder(symbol, 'limit', side, amount, price)
            else:
                order = await self.exchange.createOrder(symbol, 'market', side, amount)
            logger.info(f'Order placed: {order}')
            return order
        except Exception as e:
            logger.error(f'Failed to place order: {e}')
            raise e

    async def cancel_order(self, order_id: str, symbol: str) -> bool:
        try:
            result = await self.exchange.cancelOrder(order_id, symbol)
            logger.info(f'Order cancelled: {order_id}')
            return True
        except Exception as e:
            logger.error(f'Failed to cancel order {order_id}: {e}')
            return False

    async def get_open_orders(self, symbol: str) -> list:
        try:
            orders = await self.exchange.fetchOpenOrders(symbol)
            return orders
        except Exception as e:
            logger.error(f'Failed to get open orders for {symbol}: {e}')
            return []